Replace cassie with datastax cql driver

michaelsembwever opened the pull request:
This work is sill premature but is opened for those that want an early review on the branch.

Currently the to do list is…
- plug in a cassandra test framework (CCMBridge?) so the tests run against an embedded cassandra running (right now you need a cassandra instance running for the tests to work, the schema is self-creating so just a blank cassandra installation started up will work)
- fix the remaining tests. i do suspect CassandraSpanStore.getSpansByTraceIds(..) needs a little work still.
- double check the tests that were against Index/Storage/Aggregates still exist in some form against SpanStorage (see the commented out code in CassieSpanStoreTest)
- rename packages and classes appropriately now that things are no longer built on Cassie
- pass flags in CassandraSpanStoreFactory into the Cluster setup

The branch consists of four commits.
-  first commit is a temporary commit to comment out the redis stuff that was getting in my way, (this will be removed from the branch and PR)
- second comment is moving the hadoop aggregate stuff out of zipkin-cassandra over to zipkin-cassandra-aggregate (the latter is not in the reactor so not built – it's not compilable as it still references cassie)
- third commit is the actual code changes involved in replacing cassie with the datastax driver.
- fourth commit is removing the cassie classes, and also removing the Index/Storage/Aggregate classes.

i plan on squashing the last 3 commits into one eventually.

Otherwise the primary fasade to the datastax cql java driver is in the java class "Repository" found in the zipkin-cassandra-java subproject. (There exists rationale to have it in its own subproject). It allows automatic creation and updating of the schema, and is a good starting place for improving the datamodel. 

The datamodel does change as keyspace and table names have become all lowercase.
I've also added comments to the cql file regarding datamodel improvements that should happened afterwards (once this PR has been stabilised and is in master).

michaelsembwever committed:
TMP COMMIT
 comment out redis

michaelsembwever committed:
WIP
Replace cassie with datastax cql driver #570
 - remove all cassie classes
 - remove Storage/Index/Aggregate classes. Zipkin now only supports the SpanStorage trait.
 - remove the hadoop aggregation classes. (this needs to be rewritten from Cassie to using CqlInputFormat).
 - introduce a new java "Repository" class that provides the persistence layer using the DataStax CQL driver.
 - the Repository class sits in its own zipkin-cassandra-core subproject, a pure java dependency free artifact.
 - the schema, found in cassandra-schema-cql3.txt, is automatically created during startup.
 - minimal but incompatible changes to the schema as keyspace and table names are now all lowercase.
 - comments have been added to cassandra-schema-cql3.txt as to how the datamodel can and should be improved.

ref: https://github.com/openzipkin/zipkin/issues/458

codefromthecrypt commented:
Good progress! Here are some thoughts:
- I wouldn't block this on embedded cassandra, as that's polish. We can just use travis service/local cassandra for now (like we do today).
- Mixed feelings about the separate `-java` and `-aggregate` projects. The artifact graph is already complex, and something we've been working hard to make sense of. Can we delay this optimization? For example, add a java source tree and make the deps for `-aggregate` optional. 

I can help look at the test, but any chance you can move this branch to openzipkin? Easier for us to collaborate there.

codefromthecrypt commented:
ps.. given a choice, I'd just delete cassandra-aggregate vs move it to an incompatible peer project. I'm sure @eirslett if not others could make short work of that once the core cassandra is in.

codefromthecrypt commented:
Based on offline chat, here's my take:
- the `-java` project is already in use for a project that writes traces directly into cassandra vs via scribe etc. Maybe the name could be changed, but understand where that's coming from. Personally, I prefer `-core` suffix for library code.
- we should definitely delete the -aggregate codebase as opposed to orphaning it in a way that doesn't compile. This is part of collateral damage needed to get zipkin running again. Hopefully, @eirslett could help add a new `-aggregate` codebase after this is merged (based on git history or from scratch).

Need some folks to weigh in, as I'd like to depend on this work in order to complete other work (like removing Index/Store, and the Builder stuff).

codefromthecrypt commented on zipkin-cassandra-java/src/main/java/org/twitter/zipkin/storage/cassandra/Repository.java:
```diff
@@ -0,0 +1,749 @@
+            for (Row row : session.execute(bound).all()) {
+                long traceId = row.getBytes("key").getLong();
+                if (!spans.containsKey(traceId)) {
+                    spans.put(traceId, new ArrayList<>());
```
for JDK 7 to work, replace this with `new ArrayList<ByteBuffer>()`

michaelsembwever replied:
thanks for spotting that Adrian. 

java7 isn't specified in the build.gradle yet? or is it just my ide not picking it up maybe.

codefromthecrypt replied:
Travis builds w/JDK7 and 8. JDK7 is particularly important to retain due to
current scala version 2.10

jamescway commented:
answering @adriancole's points
- the name of the pure-java impl
  - not sure if java needs to be in the name since its self evident?  +1 to core or something based on current name scheme.
- whether it is ok to ditch cassandra-aggregates temporarily
  - :+1: , but don't know if there's people out there that might be in an awkward spot, hopefully they come on gitter and ask for help.

codefromthecrypt commented on zipkin-cassandra/src/main/scala/com/twitter/zipkin/cassandra/CassieSpanStoreFactory.scala:
```diff
@@ -15,102 +15,22 @@
-import java.net.SocketAddress
-import scala.collection.mutable.HashSet
 
 trait CassieSpanStoreFactory { self: App =>
```
rename to CassandraSpanStoreFactory

michaelsembwever replied:
+1
 (was intentionally leaving it until later in the PR lifecycle so the diff on the file remains clear)

michaelsembwever commented:
the `-core` suffix is a nice choice over `-java`.  so +1 on that.

also +1 on removing `-aggregate`. just so long enough people know where it went if questions are raised later on by those that didn't catch it here.

eirslett commented on zipkin-cassandra/src/main/scala/com/twitter/zipkin/storage/cassandra/CassandraSpanStore.scala:
```diff
@@ -226,47 +175,41 @@ class CassieSpanStore(
+        val spans = repository.getSpansByTraceIds(idBatch.toArray.map(Long.box), count)
+          .mapValues { case spans :java.util.List[ByteBuffer] => spans.asScala.map(spanCodec.decode(_).toSpan) }
+
+        traceIds.map(traceId => spans.get(traceId)).filter(_.isDefined).map(_.get).toSeq
```
Is it possible to do something like this?
`traceIds.map(traceId => spans.get(traceId).flatten)`

`.flatten` will (example given) turn `Seq(None, None, Some(1), None, Some(2))` into `Seq(1,2)` so you don't need to filter by isDefined and map on .get.

eirslett commented on zipkin-cassandra/src/main/scala/com/twitter/zipkin/storage/cassandra/CassandraSpanStore.scala:
```diff
@@ -277,71 +220,48 @@ class CassieSpanStore(
-        traces.insert(traceId, col.copy(timestamp = None, ttl = Some(ttl)))
-      }
-      traces.execute().unit
+    getSpansByTraceId(traceId).get foreach { span =>
```
`.get` is dangerous... (purists say it shouldn't be in the Scala library in the first place)
Maybe `.flatten` works here too?

eirslett replied:
Ah, it's a Future. You'll have to use

``` scala
getSpansByTraceId(traceId) map { spans => spans foreach { span =>
```

Then you won't need to return Future.Unit in the last line.

eirslett commented on zipkin-cassandra/src/main/scala/com/twitter/zipkin/storage/cassandra/CassandraSpanStore.scala:
```diff
@@ -277,71 +220,48 @@ class CassieSpanStore(
+    pool {
+      repository
+        .tracesExist(traceIds.toArray.map(Long.box))
+        .map(_.asInstanceOf[Long])
```
I think this is the same as `.map(Long.unbox)` (mentioning it since `Long.box` was used above)

eirslett commented:
Wow, this is some Great work! :grinning: :clap: :+1:

codefromthecrypt commented on zipkin-cassandra/src/test/scala/com/twitter/zipkin/storage/cassandra/CassieSpanStoreTest.scala:
```diff
@@ -15,26 +15,320 @@
+import org.scalatest.{BeforeAndAfter, FunSuite}
+
 
 class CassieSpanStoreTest extends FunSuite {
```
currently trying to work around our famous `org.apache.thrift:libthrift:0.5.0-1` vs `0.9.2` runtime compat story.

eirslett commented:
We can run zipkin-aggregate off a released version of zipkin-cassandra (with cassie). For now. But I'm pretty sure @michaelsembwever has some experience working with Cassandra taps?

codefromthecrypt commented:
Great point. Once our jars are published, we could switch the dep in the
aggregated module until the new code is ported over.

michaelsembwever commented:
rebased off master and included commits from @adriancole 's branch in https://github.com/thelastpickle/zipkin/pull/1

zipkin-cassandra-aggregate is removed.
zipkin-cassandra-java is now called zipkin-cassandra-core.

CassieSpanStoreFactory renamed to CassandraSpanStoreFactory
CassieSpanStoreSpec renamed to CassandraSpanStoreSpec

from here i'll fix up the tests and incorporate @eirslet's improvements.

eirslett commented on zipkin-cassandra/src/main/scala/com/twitter/zipkin/storage/cassandra/CassandraSpanStore.scala:
```diff
@@ -389,27 +313,33 @@ class CassieSpanStore(
+        (repository.getTraceDuration(false, traceIdSet)
+        .map { case (traceId :java.lang.Long, ts :java.lang.Long) => (traceId, ("end", ts))}
+        .toSeq))
+      .groupBy(_._1)
```
I suggest we give ._1 and ._2 names, for readability:

``` scala
.groupBy { case (whatevernameofthekey, _) => whatevernameofthekey }
```

michaelsembwever replied:
done

eirslett commented on zipkin-cassandra/src/main/scala/com/twitter/zipkin/storage/cassandra/CassandraSpanStore.scala:
```diff
@@ -389,27 +313,33 @@ class CassieSpanStore(
+        .map { case (traceId :java.lang.Long, ts :java.lang.Long) => (traceId, ("end", ts))}
+        .toSeq))
+      .groupBy(_._1)
+      .mapValues(_.map(_._2))
```
I think you can use

``` scala
.mapValues(_.values)
```

or something with .flatten? (I'm not sure if that's possible)

michaelsembwever replied:
i played around with `.mapValues(_.values)` and `.flatten` and found nothing that worked :-(

i'm open to working suggestions otherwise.

eirslett commented on zipkin-cassandra/src/main/scala/com/twitter/zipkin/storage/cassandra/CassandraSpanStore.scala:
```diff
@@ -374,10 +294,14 @@ class CassieSpanStore(
-    // TODO: endTs seems wrong here
-    idx.getRowSlice(key, Some(endTs), None, limit, Order.Reversed) map colToIndexedTraceId
+    pool {
+      (spanName match {
```
An intermediary variable could be used here, for readability?

```
val traceIds = spanName match { ... }
traceIds.map { case ... }
```

michaelsembwever replied:
i've left this. i can't think of an intuitive variable name here, hence the comment.

the closest variable name i can think of is timestampByTraceId, but i find that not intuitive enough for two lines of code.

eirslett commented on zipkin-cassandra/src/main/scala/com/twitter/zipkin/storage/cassandra/CassandraSpanStore.scala:
```diff
@@ -277,71 +220,48 @@ class CassieSpanStore(
-      else
-        Future.value(ttl.seconds)
+    pool {
+      Duration(repository.getSpanTtl(traceId), java.util.concurrent.TimeUnit.SECONDS)
```
How about `repository.getSpanTtlInSeconds(traceId)`? A longer method name to compansate for the lack of time unit?

michaelsembwever replied:
done

michaelsembwever committed:
TMP COMMIT
 comment out collector-cassandra and query-cassandra as the Store.Builder still uses (Storage,Index,Aggregate) instead of SpanStore

michaelsembwever committed:
SQUASH ME
 - improves the error logs so that we trap full cql query string with arguments

codefromthecrypt commented:
Here are some hints on how to retain support for Builder w/o having to implement Index or Storage:

there's a bridge you can use ex here: https://github.com/openzipkin/zipkin/blob/master/zipkin-collector-service/config/collector-redis.scala#L30
also here https://github.com/openzipkin/zipkin/pull/583/files#diff-1b0789c140f012e2404ae48339a50569

codefromthecrypt commented:
The following commit completes this change https://github.com/adriancole/zipkin/commit/84731aca6a4d162bce404580723be874beaef0d9

I'll raise a PR, squashing that into mck's work.

codefromthecrypt commented:
raised #585 to merge this in (removing the temp commits and squashing)
